### PR TITLE
Consider PG min resource when enqueue job resource

### PR DESCRIPTION
**Pull request by TommyLike**
Now PG min resource would be considered when calculating the node idle resource.

### CODE REVIEW

**Comment by TommyLike** [comment]
@k82cn @hzxuzhonghu please take a review

File: pkg/scheduler/actions/enqueue/enqueue.go [hunk 1]
```diff
@@ -63,6 +66,15 @@ func (enqueue *enqueueAction) Execute(ssn *framework.Session) {
 			}
 		}
 
+		//Reserve pg min resource in advance.
```
Comment (hzxuzhonghu) [inline pkg/scheduler/actions/enqueue/enqueue.go]: nit: add a white space after //

File: pkg/scheduler/actions/enqueue/enqueue.go [hunk 2]
```diff
@@ -63,6 +66,15 @@ func (enqueue *enqueueAction) Execute(ssn *framework.Session) {
...
+		if (job.PodGroup.Status.Phase == scheduling.PodGroupInqueue ||
+			job.PodGroup.Status.Phase == scheduling.PodGroupRunning) &&
+			job.PodGroup.Status.Succeeded != job.PodGroup.Spec.MinMember &&
+			job.PodGroup.Spec.MinResources != nil {
```
Comment (hzxuzhonghu) [inline pkg/scheduler/actions/enqueue/enqueue.go]: What if when one pod completed and deleted? FYI: https://github.com/volcano-sh/volcano/issues/457

Comment (TommyLike) [reply pkg/scheduler/actions/enqueue/enqueue.go]: @hzxuzhonghu what's your suggestion here?

File: pkg/scheduler/actions/enqueue/enqueue.go [hunk 3]
```diff
@@ -77,15 +89,29 @@ func (enqueue *enqueueAction) Execute(ssn *framework.Session) {
...
+	deltaRes := api.EmptyResource()
+	if pgUsedRes.Less(nodeUsedRes) {
+		deltaRes = nodeUsedRes
+	} else {
+		deltaRes = pgUsedRes
+	}
+	if deltaRes.Less(nodesIdleRes) {
+		nodesIdleRes.Sub(deltaRes)
+	} else {
+		nodesIdleRes = api.EmptyResource()
```
Comment (hzxuzhonghu) [inline pkg/scheduler/actions/enqueue/enqueue.go]: Not understand what does this do?

File: pkg/scheduler/actions/enqueue/enqueue.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pkg/scheduler/actions/enqueue/enqueue.go]: The pg is running. Why do you add its MinResources?

Comment (TommyLike) [reply pkg/scheduler/actions/enqueue/enqueue.go]: Considering all of the resources that will / already being used for pod group resources.

File: pkg/scheduler/actions/enqueue/enqueue.go (hunk 3 above)
Comment (TommyLike) [reply pkg/scheduler/actions/enqueue/enqueue.go]: use either node used resource or pg used resource that which is greater here. to perform the strict check。

File: pkg/scheduler/actions/enqueue/enqueue.go [hunk 4]
```diff
@@ -63,6 +66,15 @@ func (enqueue *enqueueAction) Execute(ssn *framework.Session) {
 			}
 		}
 
+		// Reserve pg min resource in advance.
+		if (job.PodGroup.Status.Phase == scheduling.PodGroupInqueue ||
```
Comment (k82cn) [inline pkg/scheduler/actions/enqueue/enqueue.go]: let use swithc for `job.PodGroup.Status.Phase`.

File: pkg/scheduler/actions/enqueue/enqueue.go [hunk 5]
```diff
@@ -77,15 +89,29 @@ func (enqueue *enqueueAction) Execute(ssn *framework.Session) {
 	emptyRes := api.EmptyResource()
 	nodesIdleRes := api.EmptyResource()
 	for _, node := range ssn.Nodes {
-		nodesIdleRes.Add(node.Allocatable.Clone().Multi(1.2).Sub(node.Used))
+		nodesIdleRes.Add(node.Allocatable.Clone().Multi(1.2))
```
Comment (k82cn) [inline pkg/scheduler/actions/enqueue/enqueue.go]: 1. the Clone of Allocatable is unnecessary to me.
1. multi 1.2 out of for loop.

File: pkg/scheduler/actions/enqueue/enqueue.go [hunk 6]
```diff
@@ -77,15 +89,29 @@ func (enqueue *enqueueAction) Execute(ssn *framework.Session) {
...
+		nodeUsedRes.Add(node.Used)
+	}
+
+	//deltaRes will equal to the bigger resource consumption
+	deltaRes := api.EmptyResource()
+	if pgUsedRes.Less(nodeUsedRes) {
+		deltaRes = nodeUsedRes
+	} else {
+		deltaRes = pgUsedRes
```
Comment (k82cn) [inline pkg/scheduler/actions/enqueue/enqueue.go]: I don-t think this's correct: `node.Used` resources also include the Pod out of Volcano control.

File: pkg/scheduler/actions/enqueue/enqueue.go [hunk 7]
```diff
@@ -77,15 +89,29 @@ func (enqueue *enqueueAction) Execute(ssn *framework.Session) {
...
+		nodesIdleRes = api.EmptyResource()
 	}
 
 	for {
 		if queues.Empty() {
 			break
 		}
 
-		if nodesIdleRes.Less(emptyRes) {
+		if nodesIdleRes.LessEqual(emptyRes) {
```
Comment (k82cn) [inline pkg/scheduler/actions/enqueue/enqueue.go]: why not `.IsEmpty()`?

**Comment by k82cn** [comment]
we need e2e test for this fix.

File: pkg/scheduler/actions/enqueue/enqueue.go (hunk 5 above)
Comment (TommyLike) [reply pkg/scheduler/actions/enqueue/enqueue.go]: method 'Multi' will change the object itself, Clone required if we put the logic inside of for loop

File: pkg/scheduler/actions/enqueue/enqueue.go (hunk 6 above)
Comment (TommyLike) [reply pkg/scheduler/actions/enqueue/enqueue.go]: This is the logic before my changes:
```
	for _, node := range ssn.Nodes {
		nodesIdleRes.Add(node.Allocatable.Clone().Multi(1.2).Sub(node.Used))
	}

```
what's your suggestion here?

File: pkg/scheduler/actions/enqueue/enqueue.go (hunk 4 above)
Comment (TommyLike) [reply pkg/scheduler/actions/enqueue/enqueue.go]: Phase is just part of the conditions here.

**Comment by TommyLike** [comment]
> we need e2e test for this fix.

will do

File: pkg/scheduler/actions/enqueue/enqueue.go (hunk 4 above)
Comment (k82cn) [reply pkg/scheduler/actions/enqueue/enqueue.go]: check other condition in the `case`.

**Pull request closed without merging**